runtime-rs: moving only vCPU threads into sandbox controller

gaohuatao-1 opened the pull request:
when overhead controller exists, just contrain vCPU threads
in sandbox controller

Fixes:#5760

Signed-off-by: gaohuatao <gaohuatao@bytedance.com>

studychao commented:
/test

Ji-Xinyou reviewed (approved):
LGTM! I saw the implementation of the library.

If you use cgroup v1, the tid will be added to `cgroup.procs`, leading to the entire process to be in the same cgroup. But in cgroup v2 this is not a problem.

Ji-Xinyou commented:
/test-ubuntu

Ji-Xinyou commented:
/test

YushuoEdge commented:
Hi, cgroup v2 support is in progress: https://github.com/kata-containers/kata-containers/pull/5491. In fact, we are also discussing and using real wokload to test how to better handle cgroups, https://github.com/kata-containers/kata-containers/issues/5503. 
But I think it makes sense to modify add_task_by_tgid to add_task here. Hope these will help you:)

yaoyinnan commented:
LGTM!  @gaohuatao-1. The correct situation here should be to use `add_task` to add vCPU threads to the sandbox cgroup. But there is a problem with the implementation of `add_task` in [`cgroups-rs`](https://github.com/kata-containers/cgroups-rs). In cgroup v1, it is correctly added to tasks; in cgroup v2, it is wrongly added to cgroup.procs, refer to https://github.com/kata-containers/cgroups-rs/blob/main/src/lib.rs#L358. This is because it does not take into account the thread mode in cgroup v2 itself, and does not implement related methods for using thread mode.
I modified it in https://github.com/kata-containers/cgroups-rs/pull/91, and it supports related methods of thread mode. And implemented runtime-rs support cgroup v2 in https://github.com/kata-containers/kata-containers/pull/5491. Welcome to help me to review, @gaohuatao-1, @studychao, @Ji-Xinyou, @YushuoEdge, thanks.
The PR that supports cgroup v2 still needs some review and testing to be merged, which may take a long time. So, my suggestion is that this PR can be merged first. The existing runtime-rs only supports cgroup v1. Merging it can make it work normally when `sandbox_cgroup_only = false`.

yaoyinnan commented:
/test-arm

gaohuatao-1 commented:
> LGTM! @gaohuatao-1. The correct situation here should be to use `add_task` to add vCPU threads to the sandbox cgroup. But there is a problem with the implementation of `add_task` in [`cgroups-rs`](https://github.com/kata-containers/cgroups-rs). In cgroup v1, it is correctly added to tasks; in cgroup v2, it is wrongly added to cgroup.procs, refer to https://github.com/kata-containers/cgroups-rs/blob/main/src/lib.rs#L358. This is because it does not take into account the thread mode in cgroup v2 itself, and does not implement related methods for using thread mode. I modified it in [kata-containers/cgroups-rs#91](https://github.com/kata-containers/cgroups-rs/pull/91), and it supports related methods of thread mode. And implemented runtime-rs support cgroup v2 in #5491. Welcome to help me to review, @gaohuatao-1, @studychao, @Ji-Xinyou, @YushuoEdge, thanks. The PR that supports cgroup v2 still needs some review and testing to be merged, which may take a long time. So, my suggestion is that this PR can be merged first. The existing runtime-rs only supports cgroup v1. Merging it can make it work normally when `sandbox_cgroup_only = false`.

Wonderful！
